adapters/sonobi/sonobi.go

```diff
@@ -0,0 +1,300 @@
+package sonobi
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"github.com/prebid/prebid-server/pbs"
+	"golang.org/x/net/context/ctxhttp"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"sort"
+	"strconv"
+)
+
+// SonobiAdapter - Sonobi SonobiAdapter definition
+type SonobiAdapter struct {
+	http *adapters.HTTPAdapter
+	URI  string
+}
+
+// Name returns the name fo cookie stuff
+func (a *SonobiAdapter) Name() string {
+	return "sonobi"
+}
+
+//SkipNoCookies flag for skipping no cookies...
+func (a *SonobiAdapter) SkipNoCookies() bool {
+	return false
+}
+
+// NewSonobiAdapter create a new SovrnSonobiAdapter instance
+func NewSonobiAdapter(config *adapters.HTTPAdapterConfig, endpoint string) *SonobiAdapter {
+	return NewSonobiBidder(adapters.NewHTTPAdapter(config).Client, endpoint)
+}
+
+// NewSonobiBidder Initializes the Bidder
+func NewSonobiBidder(client *http.Client, endpoint string) *SonobiAdapter {
+	a := &adapters.HTTPAdapter{Client: client}
+
+	return &SonobiAdapter{
+		http: a,
+		URI:  endpoint,
+	}
+}
+
+type sonobiParams struct {
+	TagID string `json:"TagID"`
+}
+
+// Call OpenRTB request to sonobi and parse the response into prebid server bids
+func (a *SonobiAdapter) Call(ctx context.Context, req *pbs.PBSRequest, bidder *pbs.PBSBidder) (pbs.PBSBidSlice, error) {
+	supportedMediaTypes := []pbs.MediaType{pbs.MEDIA_TYPE_BANNER}
+	sReq, err := adapters.MakeOpenRTBGeneric(req, bidder, a.Name(), supportedMediaTypes)
+	if err != nil {
+		return nil, err
+	}
+	sonobiReq := openrtb.BidRequest{
+		ID:   sReq.ID,
+		Imp:  sReq.Imp,
+		Site: sReq.Site,
+		User: sReq.User,
+		Regs: sReq.Regs,
+	}
+
+	// add tag ids to impressions
+	for i, unit := range bidder.AdUnits {
+		var params openrtb_ext.ExtImpSonobi
+		err = json.Unmarshal(unit.Params, &params)
+		if err != nil {
+			return nil, err
+		}
+
+		// Fixes some segfaults. Since this is legacy code, I'm not looking into it too deeply
+		if len(sonobiReq.Imp) <= i {
+			break
+		}
+		sonobiReq.Imp[i].TagID = params.TagID
```

Comment:
No test of the contents of params has been performed.  Is it safe to do this assignment without checking whether params.TagID is set / valid?

Reply:
Go always initializes struct data to a valid null value. So the assignment will not result in a segfault. Beyond that, the actual bidder is likely in a better position to say if a `Tag.ID` is a valid ID or not. Also the bodder paramas JSON spec below (in static) ensures that there was a `TagID` property in the `ext`.

Reply:
Okay, it makes sense to not burden PBS with validation, so given that params.TagID exists, I have no objection here.

Reply:
Aren't params validated before we reach this point by the static/bidder-params/sonobi.json spec?

Also during my test I dont ever actually hit this function when I put a debug point at the top of "Call". When is this function used? I wrote this function a few months back before the definition of an adapter changed to using MakeRequests. Only reason I left is it because its a required function in the past. Seems like its not anymore. So I'll remove Call